core/tee/tee_pobj.c

```diff
@@ -37,20 +37,21 @@ TAILQ_HEAD_INITIALIZER(tee_pobjs);
 static TEE_Result tee_pobj_check_access(uint32_t oflags, uint32_t nflags)
 {
 	/* meta is exclusive */
-	if (oflags | TEE_DATA_FLAG_ACCESS_WRITE_META ||
-	    nflags | TEE_DATA_FLAG_ACCESS_WRITE_META)
-		return TEE_ERROR_ACCESS_CONFLICT;
-
-	if (oflags | TEE_DATA_FLAG_ACCESS_READ &&
-	    !((nflags | TEE_DATA_FLAG_SHARE_READ) &&
-	      oflags | TEE_DATA_FLAG_SHARE_READ))
-		return TEE_ERROR_ACCESS_CONFLICT;
-
-	if (oflags | TEE_DATA_FLAG_ACCESS_WRITE &&
-	    !((nflags | TEE_DATA_FLAG_SHARE_WRITE) &&
-	      oflags | TEE_DATA_FLAG_SHARE_WRITE))
+	if (oflags & TEE_DATA_FLAG_ACCESS_WRITE_META)
 		return TEE_ERROR_ACCESS_CONFLICT;
 
+	/*second open/create need set flags to TEE_DATA_FLAG_SHARE_READ */
+	if (((oflags & TEE_DATA_FLAG_ACCESS_READ) ||
```

Comment:
You're right, what I have given above is not complete. But I do not think your suggestion fits as if there is no SHARE_READ at all, then it will result in an ACCESS_CONFLICT, which may be wrong if there is no ACCESS_READ at all.

I suggest the following instead (with the same tests on write, not given here):

    /* meta is exclusive */
    if (oflags & TEE_DATA_FLAG_ACCESS_WRITE_META)
        return TEE_ERROR_ACCESS_CONFLICT;

    /*
     * Excerpt of TEE Internal Core API Specification v1.1:
     * If more than one handle is opened on the same  object, and if any
     * of these object handles was opened with the flag
     * TEE_DATA_FLAG_ACCESS_READ, then all the object handles MUST have been
     * opened with the flag TEE_DATA_FLAG_SHARE_READ
     */
    if (((oflags & TEE_DATA_FLAG_ACCESS_READ) ||
         (nflags & TEE_DATA_FLAG_ACCESS_READ)) &&
        !((nflags & TEE_DATA_FLAG_SHARE_READ) &&
          (oflags & TEE_DATA_FLAG_SHARE_READ)))
        return TEE_ERROR_ACCESS_CONFLICT;
    
    /*
     * Excerpt of TEE Internal Core API Specification v1.1:
     * An object can be opened with only share flags, which locks the access 
     * to an object against a given mode.
     * An object can be opened with no flag set, which completely locks all 
     * subsequent attempts to access the object
     */
    if ((nflags & TEE_DATA_FLAG_SHARE_READ) != (oflags & TEE_DATA_FLAG_SHARE_READ))
        return TEE_ERROR_ACCESS_CONFLICT;

* The 1st test solves line 6 of the example table 5-14
* The 2nd test solves lines 1 to 5 of the example table 5-14
* The 3rd test solves lines 7 and 8 of the example table 5-14